Add docs to readme.md for running test suite in Docker

cdesch opened the pull request:
- Add docs for running the test suite in docker for enabling contributors to test their code prior to submitting
- Add `lcov` to base `Dockerfile` as a dependency so tests that run in the docker container can produce code coverage metrics

HalosGhost commented on Dockerfile:
```diff
@@ -6,13 +6,14 @@ ENV DEBIAN_FRONTEND noninteractive
-      libgmock-dev \
-      net-tools \
-      git
+    build-essential \
+    wget \
+    cmake \
+    libgtest-dev \
+    libgmock-dev \
+    net-tools \
+    lcov \
+    git
```
The purpose of the extra indent was for visual clarity that each of these lines is an argument to the command `apt install -y` rather than a new command. Additionally, by dedenting all these lines, it hides the real change (which is the addition of `lcov` to the list).

Unless there's a good reason we should remove the indentation, I think we should leave it as-is.

cdesch replied:
I believe the indentation change was a side effect of the VS Code Docker plugin auto-formatting the file on save. I think it is okay to leave the indentation as it was, although this might become a recurring issue for contributors.

cdesch replied:
I restored the indentation in this commit https://github.com/mit-dci/opencbdc-tx/pull/58/commits/a27f2100fa7da1ee4f27312440e93808a4b51b32 , although I had to disable the VS Code Docker Extension. You may run into this problem again with other contributors.

HalosGhost commented on Dockerfile:
```diff
@@ -6,13 +6,14 @@ ENV DEBIAN_FRONTEND noninteractive
+    libgtest-dev \
+    libgmock-dev \
+    net-tools \
+    lcov \
```
Allowing the scripts we provide to be run inside the docker containers is probably a good idea (if for no other reason, because of the principle of least surprise).

However, just to note, we are compatible with [`act`](https://github.com/nektos/act), so you can run the full CI suite locally (in-docker) before opening a PR.

cdesch replied:
I think this is a good idea

HalosGhost commented on README.md:
````diff
@@ -139,3 +139,45 @@ In each of the below commands, you should pass `atomizer-compose.cfg` instead of
+   ```terminal
+   # docker build . -t opencbdc-tx
+   ```
+2. Run Unit Tests
````
This will actually run the full test-suite (both unit and integration tests).

HalosGhost commented on README.md:
````diff
@@ -139,3 +139,45 @@ In each of the below commands, you should pass `atomizer-compose.cfg` instead of
+   # docker run -ti opencbdc-tx ./scripts/test.sh
+   ```
+
+Running Integration Tests
````
This is not an integration test; it's a *very* simple system test which lives outside our test suite. It's essentially just a short-cut for the walkthrough we provide above for ease-of-use.

I don't think it makes sense to document this script here; if it moves to become part of a more fully-featured system-test suite under `test/`, then I think that's a different story.

cdesch replied:
@HalosGhost Thanks. I think so too given that context. It helps to know those differences. I removed it here: https://github.com/mit-dci/opencbdc-tx/pull/58/commits/56150189dec9b6939d13f79673ea4d650cc232ea

HalosGhost reviewed (changes requested):
Thank you for the PR! There are a few changes that I think should be made.

I also have a question: this change allows `./scripts/test.sh` to be run under docker; should we go the next step and also setup `./scripts/lint.sh` to be run under docker?

If not, where's the line for what we should enable to run inside docker and what we shouldn't? If so, does it make more sense for us to document using `act` to run the full CI locally? (It's definitely a simpler process than having to manually run the containers.)

cdesch commented:
@HalosGhost I think it is a good idea to add the `clang-format` for linting.  If you're gonna have one docker image for development and running, allowing that one image to both build, test, lint and run makes it easier for new contributors to pick up the code and get started. If the docker image needs to be slimmed down because of these additional dependencies that aren't needed at runtime, I think it would be better to have a `dev` Dockerfile and a `"prod"` Dockerfile, even though the system is Production, by design 👍.

I can make a separate PR for  `clang-format` and `./scripts/lint.sh` if you would like.

cdesch reviewed (commented):
Updated PR per request

cdesch commented:
@HalosGhost I'm not sure if there is a github action for closing the review or not. Not sure if there is something else you want me to do.

HalosGhost commented:
@cdesch sorry about the delay! In the top-right of the page (in the Reviewers section), there's a button to re-request a review ([cf. step 7](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review)). That's usually my cue to take another look at a PR.

I think this PR looks pretty good now, but I think the commits should be squashed to a single commit combining the texts.

cdesch committed:
Adding Dep and  docs for running the test suite in docker

- Adding process for running unit tests
- Add `lcov` to Dockerfile for testing with `./scripts/test.sh`. `lcov` is needed for code coverage as noted in the [Contributing Docs](https://github.com/mit-dci/opencbdc-tx/blob/trunk/docs/contributing.md#testing)

Signed-off-by: Chris Desch <cdesch@gmail.com>

cdesch commented:
Thanks @HalosGhost , I squashed the commits into https://github.com/mit-dci/opencbdc-tx/commit/6c1abf0ce512d6e90ab746c21a004c450950c8b1 and performed the "Re-Review" action. That was kind of hidden up there. Thanks for guiding me.

HalosGhost reviewed (approved):
This looks good to me. Adds a dependency to the docker image so that `./scripts/test.sh` can be run in-docker, and includes a basic walkthrough for users to do so in the README.

Will merge after all checks have finished running (unless something surprising happens).

HalosGhost commented:
@cdesch happily! Welcome aboard!
